rust/plan.md

```diff
@@ -24,209 +24,243 @@ A suggested workflow:
 
-  - TODO: This subject isn't _needed_ to do the project, but formatting
-    is a good deep-dive opportunity
+**Readings**:
+ - [The Book - Macros](https://doc.rust-lang.org/book/ch19-06-macros.html)
+ - [`std::fmt`](https://doc.rust-lang.org/std/fmt/index.html)
 
+**Topics**: formatting tips, derive Debug in depth, how does `format!` work?,
+`log` demo, `slog` and structured logging, `env_logger`, APIs that take format
+strings
 
 
 
 
 ## Section 2 (File I/O)
 
-- Project: File I/O
 
-  - Task: create a persistent key/value store that can be accessed from the
-    command line
+### [Project: File I/O][p-fs]
+
+**Task**: Create a persistent key/value store that can be accessed from the
+command line
+
+**Goals**:
+
+- Handle and report errors robustly
+- Write data to disk as a write-ahead log
+- Read the state of the key/value store from disk
+- Use serde for serialization
+- Use file I/O APIs to binary search
+
+**Topics**: `failure` crate, `std::net::fs`, `Read` / `Write` traits,
+serde
 
-  - Goals:
-    - Handle and report errors robustly
-    - Accept command line arguments
-    - Write data to disk as a write-ahead log
-    - Read the state of the key/value store from disk
-    - Use serde for serialization
-    - Maintain consistency after crashes or data corruption?
-    - Use file I/O APIs to binary search
+**Extensions**: range queries, store data using bitcast algo?
 
-  - Extensions: range queries, convert full WAL to binary-searchable file
+### [Lesson: Proper error handling][t-errors] ([slides][s-errors])
 
-- Lesson: Proper error handling
+**Readings**:
 
-  - Readings:
-    - [The Book - Error Handling](https://doc.rust-lang.org/book/ch09-00-error-handling.html)
-    - [Rust Error Handling](http://blog.burntsushi.net/rust-error-handling/)
-    - [The `failure` book](https://rust-lang-nursery.github.io/failure/)
+- [The Book - Error Handling](https://doc.rust-lang.org/book/ch09-00-error-handling.html)
+- [Rust Error Handling](http://blog.burntsushi.net/rust-error-handling/)
+- [The `failure` book](https://rust-lang-nursery.github.io/failure/)
 
-  - Topics: TODO, `fn main() -> Result`
+**Topics**: error patterns, `failure` crate, `fn main() -> Result`, `panic!` and
+unwinding
 
-- Lesson: Collections and iterators
+### [Lesson: Collections and iterators][t-coll] ([slides][s-coll])
 
 
 
 
 ## Section 3 (Networking)
 
-- Project: Networking
 
-  - Task: create a single-threaded, persistent key/value store server and client
-    with synchronous networking over the HTTP and GRPC protocols.
+### [Project: Networking][p-net]
 
-- Lesson: Basic network APIs
+**Task**: Create a single-threaded, persistent key/value store server and client
+with synchronous networking over the HTTP protocol.
 
-  - Topics: std networking, TCP vs UDP?, reqwest, blocking HTTP serving w/ Iron/Hyper?
+**Goals**:
 
-- Lesson: Build-time Rust
+- Use hyper for synchronous networking
```

Comment:
To my knowledge, as HTTP servers, the latest hyper and iron are all asynchronous and run over the tokio runtime. Maybe we can consider use `http` crate to build a synchronous HTTP server ourselves?

Reply:
My fault. Only at the master branch iron uses hyper 0.12 but the released 0.6.0 version uses the synchronous hyper 0.10.

Reply:
Then, how will we make it single threaded first and change it to multi-threaded in the next project? By using hyper 0.10 or iron, that's simply to set the thread num of the pool to 1 first and then make it larger. This does not make much sense.

Reply:
I was under the impression hyper still offered a synchronous API, but it does not seem to.

Building a synchronous HTTP server could be interesting, requiring direct use of the std networking APIs. It's also demystifying, shows that HTTP is approachable. It's not all that practical, since most people should not implement HTTP themselves.

Another solution would be to use gRPC instead of HTTP.

I'm going to leave this an open question.